### PR TITLE
Make Obj.dup use a new primitive, %obj_dup

### PR DESCRIPTION
**Pull request by mshinwell**
Instead of calling `caml_obj_dup` from user code in order to copy a locally-allocated value onto the heap, so that it can escape, we could use a compiler primitive in some cases (e.g. for boxed numbers, where no recursion is necessary).

This allows a pattern where constructor functions always return local values that are typically globalized shortly afterwards, but where only a single (heap) allocation occurs (instead of a local allocation then a heap allocation).  For example after this PR, the function `f` below simplifies under Flambda 2 to a single heap allocation:
```
module Foo : sig
  type t = private float

  val create : float -> local_ t
  val globalize : local_ t -> t
end = struct
  type t = float

  let[@inline] create i = local_ (i *. 2.)

  external globalize : local_ 'a -> 'a = "%globalize"
end

module Use : sig
  val f : float -> Foo.t
end = struct
  let f i =
    let local_ foo = Foo.create i in
    Foo.globalize foo [@nontail]
end
```
We might also want something like this for strings, which will require some extra handling in Flambda 2 (e.g. to track the results of basic string operations such as concatenation).

A ppx preprocessor can be used on top of this to implement globalization for the recursive cases of variants, records, etc.  The existing Flambda 1 and 2 code is sufficient to simplify those cases, e.g. in this example `f` simplifies to a single heap allocation:
```
module Foo : sig
  type t = private Foo of int

  val create : int -> local_ t
  val globalize : local_ t -> unit -> t
end = struct
  type t = Foo of int

  let[@inline] create i = local_ (Foo i)

  external globalize_int : local_ int -> int = "%identity"

  let[@inline] globalize (local_ t) () =
    match t with
    | Foo i -> Foo (globalize_int i)
end

module Use : sig
  val f : int -> Foo.t
end = struct
  let f i =
    let local_ foo = Foo.create i in
    Foo.globalize foo () [@nontail]
end
```
(Note that the `unit` argument to `globalize` shouldn't be needed; we seem to have a bug whereby the `nlocal` argument to the `Curried` constructor is either wrong or being misinterpreted under Flambda 2 -- I'm discussing with @stedolan .)

Based on #809.

### CODE REVIEW

**Comment by mshinwell** [comment]
I spoke with @stedolan about this; here is a revised version.  The new primitive is now `%obj_dup` and this is used for the implementation of `Obj.dup`.  The only time that the semantics of this would need to differ from a "globalize" primitive is for mutable values (for globalize, it would not be necessary to perform a new allocation if the provided block was already on the heap, whereas for `Obj.dup` a copy must be made).  Given we don't currently have plans for tracking of the contents of mutable values in Flambda 2 (the contents tracking being required for examples such as the above), it seems fine to just have a single `%obj_dup` primitive for now.

This patch includes a bootstrap because of the use of `%obj_dup` in the stdlib.

File: ocaml/lambda/lambda.ml [hunk 1]
```diff
@@ -211,6 +211,8 @@ type primitive =
   | Popaque
   (* Statically-defined probes *)
   | Pprobe_is_enabled of { name: string }
+  (* Local allocations *)
```
Comment (lthls) [inline ocaml/lambda/lambda.ml:214]: It's not really about local allocations. Could you think of a better comment ?

File: ocaml/lambda/lambda.mli [hunk 2]
```diff
@@ -181,6 +181,8 @@ type primitive =
   | Popaque
   (* Statically-defined probes *)
   | Pprobe_is_enabled of { name: string }
+  (* Local allocations *)
```
Comment (lthls) [inline ocaml/lambda/lambda.mli:184]: Same as in `lambda.ml`

File: ocaml/lambda/printlambda.ml [hunk 3]
```diff
@@ -439,6 +439,7 @@ let primitive ppf = function
   | Pint_as_pointer -> fprintf ppf "int_as_pointer"
   | Popaque -> fprintf ppf "opaque"
   | Pprobe_is_enabled {name} -> fprintf ppf "probe_is_enabled[%s]" name
+  | Pobj_dup -> fprintf ppf "globalize"
```
Comment (lthls) [inline ocaml/lambda/printlambda.ml:442]: Leftover reference to `globalize`

File: middle_end/flambda2/types/flambda2_types.mli [hunk 4]
```diff
@@ -578,15 +578,15 @@ val prove_variant_like :
...
 (** If [ty] is known to represent a boxed number or a tagged integer,
-    [prove_is_a_boxed_number env ty] is [Proved kind]. [kind] is the kind of the
-    unboxed number.
+    [prove_is_a_boxed_number env ty] is [Proved (alloc_mode,kind,contents_ty)].
+    [kind] is the kind of the unboxed number.
 
     If [ty] is known to represent something of kind value that is not a number
     [prove_is_a_boxed_number env ty] is [Invalid].
 
     Otherwise it is [Unknown] or [Wrong_kind] when [ty] is not of kind value. *)
```
Comment (lthls) [inline middle_end/flambda2/types/flambda2_types.mli:587]: Remind me at some point to make a pass on this file to fix the comments (they haven't been updated in the Provers overhaul PR)

**Review by lthls — APPROVED** [review]
I've left some minor comments, apart from that everything is fine.

File: middle_end/flambda2/types/flambda2_types.mli (hunk 4 above)
Comment (mshinwell) [reply middle_end/flambda2/types/flambda2_types.mli:587]: Have made a note.

File: ocaml/lambda/lambda.ml (hunk 1 above)
Comment (mshinwell) [reply ocaml/lambda/lambda.ml:214]: Fixed

File: ocaml/lambda/lambda.mli (hunk 2 above)
Comment (mshinwell) [reply ocaml/lambda/lambda.mli:184]: Fixed

File: ocaml/lambda/printlambda.ml (hunk 3 above)
Comment (mshinwell) [reply ocaml/lambda/printlambda.ml:442]: Fixed

**Review by lthls — APPROVED** [review]

**Pull request merged**